### PR TITLE
v0.7.3 Add MTP support for deepseek

### PR DESCRIPTION
**Pull request by mengwei805**
### What this PR does / why we need it?
support MTP.

### Does this PR introduce _any_ user-facing change?
can use num_speculative_tokens=1 turn on MTP in deepseekv3/r1 model.

### How was this patch tested?
tests/spec_decode/e2e/test_mtp_correctness.py  passed.

### CODE REVIEW

File: vllm_ascend/ops/rejection_sampler.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#
```
Comment (wangxiyuan) [inline vllm_ascend/ops/rejection_sampler.py]: this change is `patch`, not `ops`, right?

File: vllm_ascend/ops/rotary_embedding.py [hunk 2]
```diff
@@ -88,5 +89,39 @@ def rope_deepseek_forward_oot(
...
+    freqs = torch.einsum("i,j -> ij", t, inv_freq)
+    cos = (freqs.cos() * self.mscale)
+    sin = (freqs.sin() * self.mscale)
+    cache = torch.cat((cos, sin), dim=-1)
+    return cache
+
+
 RotaryEmbedding.forward_oot = rope_forward_oot
 DeepseekScalingRotaryEmbedding.forward = rope_deepseek_forward_oot
+# v0.7.4 do not need. or `from torch_npu.contrib import transfer_to_npu` also cann fix it.
```
Comment (wangxiyuan) [inline vllm_ascend/ops/rotary_embedding.py]: Ditto, move the change to patch model

File: vllm_ascend/spec_decode/metrics.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
+#
```
Comment (wangxiyuan) [inline vllm_ascend/spec_decode/metrics.py]: ditto

File: vllm_ascend/ops/rejection_sampler.py (hunk 1 above)
Comment (mengwei805) [reply vllm_ascend/ops/rejection_sampler.py]: you are right, i will move it to `patch` module

File: vllm_ascend/ops/rotary_embedding.py (hunk 2 above)
Comment (mengwei805) [reply vllm_ascend/ops/rotary_embedding.py]: The changes here have been made in https://github.com/vllm-project/vllm-ascend/pull/228 by `from torch_npu.contrib import transfer_to_npu` in worker.py. I will remove this change in my PR.

File: vllm_ascend/spec_decode/metrics.py (hunk 3 above)
Comment (mengwei805) [reply vllm_ascend/spec_decode/metrics.py]: i will move it to `patch` module

File: vllm_ascend/attention.py [hunk 4]
```diff
@@ -806,11 +806,10 @@ def forward(
...
             assert kv_cache is not None
-            attn_output = torch.empty(num_tokens,
-                                      self.num_heads,
-                                      self.kv_lora_rank,
-                                      dtype=query.dtype,
-                                      device="npu")
+            attn_output = torch.randn(
+                [num_tokens, self.num_heads, self.kv_lora_rank],
+                dtype=query.dtype,
+                device="npu")
```
Comment (MengqingCao) [inline vllm_ascend/attention.py]: Since the rank of device maybe specified by users, we recommand to modify as the following:
```suggestion
                device=query.device)
```

File: vllm_ascend/patch/rejection_sampler.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+    if num_samples > 1:
+        # This is equivalent to torch.repeat_interleaved (which also
+        # forces a GPU<->CPU sync).
+        probs = probs[:, None, :].expand(probs.shape[0], num_samples,
+                                         probs.shape[1]).contiguous().view(
+                                             -1, probs.shape[1])
+    q = torch.empty_like(probs)
+    if not seeded_seqs:
+        q = -torch.log(torch.rand_like(q))
```
Comment (MengqingCao) [inline vllm_ascend/patch/rejection_sampler.py]: Is this fixes the accurancy issue when `SamplingParm.seed` is not set?

**Comment by wangxiyuan** [comment]
need a rebase

File: vllm_ascend/patch/rejection_sampler.py (hunk 5 above)
Comment (mengwei805) [reply vllm_ascend/patch/rejection_sampler.py]: The `exponential_` operator in the current torch_npu has precision issues. Offline multi-batch inference is the answer to the first question, which is incorrect. I replaced it with an equivalent implementation, which has been tested.

File: vllm_ascend/attention.py (hunk 4 above)
Comment (mengwei805) [reply vllm_ascend/attention.py]: Considering that the vllm-ascend repository should only be used on npu devices? I wrote it as a fixed npu. Maybe you are right. I will modify this place.

**Review by wangxiyuan — APPROVED** [review]
rename patch files to `patch_xxx.py` if need another update, otherwise it can be done in the following PR.

File: vllm_ascend/patch/patch_metrics.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    aggregate_metrics_ready.record(self._copy_stream)
+
+    return aggregate_metrics_ready
+
+
+AsyncMetricsCollector.__init__ = __npu_async_metrics_collector_init__
+AsyncMetricsCollector.init_gpu_tensors = init_gpu_tensors
+AsyncMetricsCollector.init_tensors = init_tensors
+AsyncMetricsCollector.maybe_collect_rejsample_metrics = maybe_collect_rejsample_metrics
+AsyncMetricsCollector._copy_rejsample_metrics_async = _copy_rejsample_metrics_async
```
Comment (wangxiyuan) [inline vllm_ascend/patch/patch_metrics.py]: IIUC, this patch only changed `cuda` to `npu`, so base on this import https://github.com/vllm-project/vllm-ascend/blob/v0.7.3-dev/vllm_ascend/worker/worker.py#L70  does this patch still need?

File: vllm_ascend/patch/rejection_sampler.py (hunk 5 above)
Comment (wangxiyuan) [reply vllm_ascend/patch/rejection_sampler.py]: Thanks for explainatin. please add a note here

File: vllm_ascend/patch/patch_metrics.py (hunk 6 above)
Comment (mengwei805) [reply vllm_ascend/patch/patch_metrics.py]: Because of the change you mentioned, the two functions `init_gpu_tensors` and `_copy_rejsample_metrics_async` do not need to be patched. However, the other two functions still need to be patched in version 0.7.3 because:
1. init_tensors: Modified the hard-coded `cuda` judgment logic to `npu`;
2. maybe_collect_rejsample_metrics: Removed the check for `current_platform.is_cuda_alike()`

I will modify it as soon as possible.

**Review by wangxiyuan — APPROVED** [review]
